database/postgresconn/postgresconn.go

```diff
@@ -611,11 +606,70 @@ AND format_type(a.atttypid, a.atttypmod) = 'bigint'`, tableName, columnName)
 
 // SetFailed set the current migration to failed and record the failure in the database
 func (p *Postgres) SetFailed(version int, err error) error {
-	ctx := context.Background()
 	stmt := fmt.Sprintf(`UPDATE %q.%q SET info = $1 where version = $2`,
 		p.config.migrationsSchemaName, p.config.migrationsTableName)
-	if _, err := p.conn.ExecContext(ctx, stmt, fmt.Sprintf("%+v", err), version); err != nil {
+	if _, err := p.conn.ExecContext(p.context(), stmt, fmt.Sprintf("%+v", err), version); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Begin begins transaction
+func (p *Postgres) Begin() error {
+	if p.tx != nil {
+		return fmt.Errorf("transaction already started")
+	}
+
+	p.ctx = context.Background()
+	tx, err := p.conn.BeginTx(p.ctx, nil)
+	if err != nil {
 		return err
 	}
+	// capture tx so we know a transaction was started
+	p.tx = tx
 	return nil
 }
+
+// Commit commits in progress transaction
+func (p *Postgres) Commit() error {
+	if p.tx == nil {
+		return fmt.Errorf("no transaction in progress")
+	}
+
+	defer func() {
+		// reset p.tx so a new transaction can be started
+		p.tx = nil
+	}()
+
+	if err := p.tx.Commit(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Rollback rolls back in progress transaction
+func (p *Postgres) Rollback() error {
+	if p.tx == nil {
+		return fmt.Errorf("no transaction in progress")
+	}
+
+	defer func() {
+		// reset p.tx so a new transaction can be started
+		p.tx = nil
+	}()
+
+	if err := p.tx.Rollback(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Context returns the a context that was created when transactions begins, or a
+// background context if no context is set(no transaction started).
+func (p *Postgres) context() context.Context {
+	if p.ctx == nil {
+		p.ctx = context.Background()
```

Comment:
Is p.ctx ever anything but the background context?

Reply:
yes, it gets wrapped if a client has a configured timeout
```
	ctx, cancel = context.WithTimeout(p.ctx, p.config.StatementTimeout)
```

Reply:
Got it, thanks